wagtail/admin/views/generic/chooser.py

```diff
@@ -133,10 +145,84 @@ def render_to_response(self):
         raise NotImplementedError()
 
 
-class ChooseViewMixin:
+class CreationFormMixin:
+    """
+    Provides a form class for creating new objects
+    """
+
+    creation_form_class = None
+    form_fields = None
+    exclude_form_fields = None
+    creation_form_template_name = "wagtailadmin/generic/chooser/creation_form.html"
+    create_action_label = _("Create")
+    create_action_clicked_label = None
+    create_url_name = None
+    permission_policy = None
+
+    def get_permission_policy(self):
+        if self.permission_policy:
+            return self.permission_policy
+        elif self.model:
+            return ModelPermissionPolicy(self.model)
+        else:
+            return BlanketPermissionPolicy(None)
+
+    def can_create(self):
+        return self.get_permission_policy().user_has_permission(
+            self.request.user, "add"
+        )
+
+    def get_creation_form_class(self):
+        if self.creation_form_class:
+            return self.creation_form_class
+        elif self.form_fields is not None or self.exclude_form_fields is not None:
+            return modelform_factory(
+                self.model, fields=self.form_fields, exclude=self.exclude_form_fields
```

Comment:
question: in `get_permission_policy` there is a check on `self.model`. Will there ever be a case where we get to this and there is no model? and this account for it. If not, is the logic in `get_permission_policy` overkill?

Reply:
Eventually we want to support choosers that work with things other than models (e.g. DRF API endpoints). My strategy at the moment is to defer thinking about that for as long as possible, design the choosers around the primary use-case (models), and then try to make the minimum necessary changes to accommodate DRF datasources (which might well involve inventing a QuerySet-like wrapper around DRF - I think part of the reason the wagtail-generic-chooser package got overcomplicated is that it was trying to accommodate multiple styles of query at once).

`get_permission_policy` is an exception to my "leave it for later" rule, because I had to make an up-front design decision about the chooser API: does it default to "allow everything", or to Django's stock permission mechanism? I chose the latter, but only if we actually have a model to base the permissions off.

In the case of the creation form, passing `form_fields` / `exclude_form_fields` only makes sense for model forms - for non-model-based forms, we'd expect developers to define a form explicitly as a `forms.Form` subclass and set that as `creation_form_class`. (It would be sensible for us to catch the case where `form_fields` is passed without setting a model, and raise an ImproperlyConfigured exception - I may well do that when I get round to properly tackling non-model choosers!)